msbuild/Xamarin.MacDev.Tasks/Decompress.cs

```diff
@@ -0,0 +1,220 @@
+			// it on Windows. It's also possible to set the XAMARIN_USE_SYSTEM_IO_COMPRESSION=1 environment
+			// variable to force using System.IO.Compression on !Windows, which is particularly useful when
+			// testing the System.IO.Compression implementation locally (with the caveat that symlinks won't
+			// be extracted).
+
+			bool rv;
+			if (Environment.OSVersion.Platform == PlatformID.Win32NT) {
+				rv = TryDecompressUsingSystemIOCompression (log, zip, resource, decompressionDir);
+			} else if (!string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("XAMARIN_USE_SYSTEM_IO_COMPRESSION"))) {
+				rv = TryDecompressUsingSystemIOCompression (log, zip, resource, decompressionDir);
+			} else {
+				rv = TryDecompressUsingUnzip (log, zip, resource, decompressionDir);
+			}
+
+			if (rv) {
+				Directory.CreateDirectory (Path.GetDirectoryName (stampFile));
+				using var touched = new FileStream (stampFile, FileMode.Create, FileAccess.Write);
+				createdFiles.Add (stampFile);
+			}
+
+			if (File.Exists (decompressedResource)) {
+				createdFiles.Add (decompressedResource);
+			} else if (Directory.Exists (decompressedResource)) {
+				createdFiles.AddRange (Directory.GetFiles (decompressedResource, "*", SearchOption.AllDirectories));
+			}
+
+			return rv;
+		}
+
+		// The dir separator character in zip files is always "/", even on Windows
+		const char zipDirectorySeparator = '/';
+
+		static bool TryDecompressUsingUnzip (TaskLoggingHelper log, string zip, string resource, string decompressionDir)
+		{
+			var archive = ZipFile.OpenRead (zip);
+			resource = resource.Replace ('\\', zipDirectorySeparator);
+			var entry = archive.GetEntry (resource);
+			if (entry is null) {
+				entry = archive.GetEntry (resource + zipDirectorySeparator);
+				if (entry is null) {
+					log.LogError (MSBStrings.E7112 /* Could not find the file or directory '{0}' in the zip file '{1}'. */, resource, zip);
+					return false;
+				}
+			}
+
+			var zipPattern = entry.FullName;
+			if (zipPattern.Length > 0 && zipPattern [zipPattern.Length - 1] == zipDirectorySeparator) {
+				zipPattern += "*";
+			}
+
+			var args = new string [] {
+				"-u", "-o",
+				"-d", decompressionDir,
+				zip,
+				zipPattern,
+			};
+			var rv = XamarinTask.ExecuteAsync (log, "unzip", args).Result;
+			return rv.ExitCode == 0;
+		}
+
+		static bool TryDecompressUsingSystemIOCompression (TaskLoggingHelper log, string zip, string resource, string decompressionDir)
+		{
+			var rv = true;
+
+			// canonicalize input
+			resource = resource.TrimEnd ('/', '\\');
+			resource = resource.Replace ('\\', zipDirectorySeparator);
+			var resourceAsDir = resource + zipDirectorySeparator;
+
+			var archive = ZipFile.OpenRead (zip);
+			foreach (var entry in archive.Entries) {
+				var entryPath = entry.FullName;
+				if (entryPath.Length == 0)
+					continue;
+
+				if (entryPath.StartsWith (resourceAsDir, StringComparison.Ordinal)) {
```

Comment:
Similar question here, is it written in this way instead of using a single if statement to make it easier to read?

Reply:
Yes, they're both written this way to make them easier to read.